### ISSUE TITLE
[Enhancement] Starting a new experimental branch

### ISSUE THREAD

**Issue by Larpoux** [dooboolab/flutter_sound#343]
@bsutton : 

I agree that starting a new experimental branch is probably the best trade-off/compromise that we can find together. I am going to rename the `dev` branch as `experimental`.

It will be possible to publish this branch as "5.0.0-experimental.1" if @hyochan agrees (he is the project leader).

The major problem that I see, is that this experimental branch looks terribly like a fork.
I am going to release next week a new Flutter Sound version with new codecs. Those codecs will be missing in the experimental branch. In the other direction, the experimental branch will probably fix some problems that we actually have in the master branch.

And more the time will pass, more we will diverge. This is what is called a fork. And a fork is not something that I like very much. Specially when we are 3 developers (@salvatore373 , you and me) and we want to work on 3 different forks.
@salvatore373 told me that he wants that I publish its branch as BETA, and just wait for a feedback from our users. If I don't do that, he will publish his own fork. I tried to explain him that we must work together to improve Flutter Sound and not split our force. Actually you are requesting exactly the same thing : publish a new fork called "experimental".

I see 3 forks when we are 3 developers as a failure of my anarchist/libertarian ideas.
I believe in anarchy. I really think that it is possible to develop great things without a boss to drive us. I was an activist in a political anarchist movement. We was just a few hundreds. Nothing. We was much too small to count in the political arena. But because we had no leaders, we always argue together, and every two years we had a new split in two new movements. This was failures. But I still continue to believe in anarchy. I believe that we can combine our forces without any leaders. 

What can we do to work together instead of forking ?

**Comment by bsutton** [comment]
@Larpoux 
whether we fork now or not is in your hands.

The simplest way to stop the fork is to stop all new features in the 4.x branch.

We put 4.x into maintenance only mode.

If users want the new features of the 5.x branch then they need move to 5.x.

This will have the following benefits:

1) A stable branch in 4.x 
2) This will generated the user feedback that we need to resolve any issues.
3) free up additional resources to test 5.x

If you continue adding new features to 4.x then we will have a fork. 
If you stop adding new features to 4.x then we won't end up with a fork.

Here is what I believe will give the best long term result for users of flutter sound.
1) cease all new development on 4.x - bug fixes only
2) develop a plugin architecture for adding new codes ( I will raise an issue for this).
3) start writing unit test for 5.x. This will overcome your concerns that 5.x is not reliable and help create a stable build going forward.
4) push out a alpha of 5.x asking for user feedback.

The project really needs both of us working together.
Its large and complex and really needs multiple skill sets.

The 5.x branch is needed as it fixes some really fundamental issues.
The above plan gives users stability and moves the project forward.

Anarchy only works if we listen to each other.

I've heard that you need stable low risk updates.
I'm saying that we need to re-architecture flutter sound if we want it to be stable with the new features we need to support.

The above plan give us both.

Its up to you now.
You need to make a management decision :)

Do we fork or not?

**Comment by bsutton** [comment]
Just thought it might be worth highlighting what our user will get if we move forward with the 5.0 branch.

* SoundPlayerUI widget
* SoundRecorderUI widget
* RecordingPlaybackController widget (pairs a recording and player widget).
* QuickPlay for short audio playback.
* consistent api (e.g. consistent usage of Tracks, naming conventions...)
* Fixes for jank when starting a player
* Fixes for a no. of crashes due to race conditions during initialisation
* Support for app pause/resume
* Stability of the api as apposed to the constant changes that are occurring now.
* Fixes for Future handling (causing more jank).
* Improved api documentation.
* Improved streaming model for duration/db tracking
* Cleaner public api - I've work to hide any internal methods.
* cleaner permission model.
* a path to a cleaner codec support model

Without 5.0 we loose most of the above and they will be a long time coming.

**Comment by Larpoux** [comment]
> whether we fork now or not is in your hands.
> 
> The simplest way to stop the fork is to stop all new features in the 4.x branch.
> If you continue adding new features to 4.x then we will have a fork.


I will say exactly the opposite : 

Whether we fork or not is in your hand. If you continue to develop a new project instead of adding changes in the master branch of Flutter Sound we will have a fork.

**Comment by bsutton** [comment]
I came to this project because my company evaluated flutter sound and came
to the conclusion that the current api had some serious flaws.

The flaws are serious enough that our team's opinion was that it needed to
be re-architectured in order to resolve those issues.
The work I've done has been to implement that re-architecture.

The 4.0 branch won't meet our internal standards and is unlikely to any
time soon.
It is also clear that with the current 4.0 approach that the api will
continue to change.
We also need the new features that I've introduced into 5.0. Those features
won't be viable any time soon in 4.0 due to its serious deficiencies.

I've laid out a compromise plan that meets both our objectives.
1) continuing stability of the 4.0 branch in a maintenance only mode
2) moving forward with 5.0 where all new features are developed.

You say you don't want to be the boss yet in a partnership there is
compromise.
I don't feel that I'm seeing much compromise from your end.

My work won't allow me to spend more time on this project unless we see a
viable way forward.
Long term support of 4.0 (beyond maintenance) is not seen as viable.

If we can find a compromise you get increased resources on the project.
Without a compromise you are back to doing by yourself. If you want to be
the boss then I guess that will work for you.

I hope we can find a compromise.

Brett

**Comment by Larpoux** [comment]
My own personal thinking is that 4.0 is a LTS version. This version is stable enough to be maintained in the long term. And I do not expect that all our users will switch to 5.0 today. They wants a stable and maintained Flutter Sound version.

For me, 5.0 is the `master` branch version. This version has several API changes, but has continuity with the Flutter Sound project. This is definitely not a new project.
I have begun to write a new API documentation which will be our(my?) functional specifications.
This rewriting of the doc is not finished (doc of the Recorder module is not finished).
but you can have a [previous look to this](https://github.com/dooboolab/flutter_sound/blob/master/README.md)

@salvatore373 did a very good job with his stuff. Unfortunately he did want to push his work inside Flutter Sound and I had to spend more than 2 weeks to implement Track Player inside Flutter Sound.

Actually I think that we *MUST* do the same thing with your fork : we will have to spend several weeks to integrate your stuff inside an already existing project. I really do not want a new project, and there are many things very good in the `dev` branch that I want to integrate inside Flutter Sound.

**Comment by bsutton** [comment]
Making 4.0 an LTS makes a lot of sense and is essentially what I'm
proposing.

Integrating pieces of the dev branch back into master is not going to work
and won't be an acceptable solution from my work's perspective.

To be honest the 4.0 code base is a mess as is the api.
If we try to integrate my dev branch into 4.0 we are just going to get a
bigger mess and the api will need to keep changing.

The bottom line is that I can continue working with you if we freeze the
4.0 branch as the LTS and make the dev branch the new master.

It will take some time for existing users to start moving to 5.0 but new
users will probably want the new features and therefore be happy to move
forward with us.

**Comment by Larpoux** [comment]
Do not forget that I am not the project leader. @hyochan is the project leader, not me. I do not have the power to veto or approve your `dev` fork.

### 1. @hyochan decide to replace Flutter Sound by your `dev` fork : 

I must confess that I would be very glad of this choice. I am currently spending too much time on Flutter Sound and my customer is actually thinking that I do not work enough on Dyglot Korean (my own project). So, if someone else can take care of the baby, I would be very happy. This would be a good choice for me, but I think a bad choice for Flutter Sound

### 2. @hyochan decide to continue the development of the Flutter Sound `master` branch.

Then, I would like to port some of the great features of your `dev` fork to Flutter Sound. With your help, I hope. For example the UI widgets. But the problem is that it will be time consuming, and spare time is something that I do not have

### 3. Your company wants to control Flutter Sound.

This is completely impossible. Flutter Sound is a free project and cannot be controlled by anybody. Specially by a private company. If your company wants to control something, the only possibility is to control a Flutter Sound fork. But do not forget that Flutter Sound is LGPL. Your company`s fork will must be published under a LGPL license, too.

**Comment by bsutton** [comment]
We aren't interested in controlling flutter sound.

Our objective is to ensure that the project is reliable and well maintained.

If we can settle in my proposed compromise then we will have more resources
available to work on the project.

That has to be better for flutter.

Put the plan I've suggested to @hyochan and let's set what he says.

**Comment by Larpoux** [comment]
Actually, the Flutter Sound project is stalled for me.
I am not going to do anything until this Management Issue is settled.

But our users are currently waiting for us, and there is urgency to fix things.

**Comment by bsutton** [comment]
Have you heard anything from the boss :)

**Comment by Larpoux** [comment]
I do not have a direct hot-line with him :-D 😄

**Comment by bsutton** [comment]
This is an open source project, if it's urgent then they might need to
create some pr's of their own:)

**Comment by bsutton** [comment]
I thought you might have exchanged email addresses.

**Comment by Larpoux** [comment]
Yes. But actually (unfortunately) free project users act and think as customers/clients and just want a free (as in a beer) product ready to be used. 👎

**Comment by bsutton** [comment]
You are right.

But then they don't have the right to demand urgency.

I believe that if you use open source that you should contribute something
back.
